fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/metadata/MetaDataEvolutionValidator.java

```diff
@@ -0,0 +1,906 @@
+            throw new MetaDataException("index name changed",
+                    LogMessageKeys.OLD_INDEX_NAME, oldIndex.getName(),
+                    LogMessageKeys.NEW_INDEX_NAME, newIndex.getName());
+        }
+        if (oldIndex.getAddedVersion() != newIndex.getAddedVersion()) {
+            throw new MetaDataException("new index added version does not match old index added version",
+                    LogMessageKeys.INDEX_NAME, newIndex.getName(),
+                    LogMessageKeys.OLD_VERSION, oldIndex.getAddedVersion(),
+                    LogMessageKeys.NEW_VERSION, newIndex.getAddedVersion());
+        }
+        if (oldIndex.getLastModifiedVersion() > newIndex.getLastModifiedVersion()) {
+            throw new MetaDataException("old index has last-modified version newer than new index",
+                    LogMessageKeys.INDEX_NAME, newIndex.getName(),
+                    LogMessageKeys.OLD_VERSION, oldIndex.getLastModifiedVersion(),
+                    LogMessageKeys.NEW_VERSION, newIndex.getLastModifiedVersion());
+        }
+        if (!allowIndexRebuilds && oldIndex.getLastModifiedVersion() != newIndex.getLastModifiedVersion()) {
+            throw new MetaDataException("last modified version of index changed",
+                    LogMessageKeys.INDEX_NAME, newIndex.getName(),
+                    LogMessageKeys.OLD_VERSION, oldIndex.getLastModifiedVersion(),
+                    LogMessageKeys.NEW_VERSION, newIndex.getLastModifiedVersion());
+        }
+        if (allowIndexRebuilds && oldIndex.getLastModifiedVersion() < newIndex.getLastModifiedVersion()) {
+            // All subsequent checks verify that the index has not been modified in a way that
+            // requires rebuilding. This is always an invalid modification if the last modified version
+            // has not changed, but it is allowed if the allowIndexRebuilds option has been set to true.
+            return;
+        }
+        if (!oldIndex.getType().equals(newIndex.getType())) {
+            throw new MetaDataException("index type changed",
+                    LogMessageKeys.INDEX_NAME, newIndex.getName(),
+                    LogMessageKeys.OLD_INDEX_TYPE, oldIndex.getType(),
+                    LogMessageKeys.NEW_INDEX_TYPE, newIndex.getType());
+        }
+        if (!oldIndex.getRootExpression().equals(newIndex.getRootExpression())) {
+            throw new MetaDataException("index key expression changed",
+                    LogMessageKeys.INDEX_NAME, newIndex.getName(),
+                    LogMessageKeys.OLD_KEY_EXPRESSION, oldIndex.getRootExpression(),
+                    LogMessageKeys.NEW_KEY_EXPRESSION, newIndex.getRootExpression());
+        }
+        // All of the record types associated with the old index should be contained in the new index.
+        // The new index might have new record types, but they need to be from record types associated
+        // with newer record types than the old meta-data.
```

Comment:
I think this could be clearer. Maybe "additional" record types and maybe strike "from ... with."